File: 33-SearchInRotatedSortedArray/33-SearchInRotatedSortedArray.cpp
// Last updated: 3/18/2026, 3:02:44 PM
class Solution {
public:
    int findPivot(vector<int>& nums) {
        int l = 0, r = nums.size() - 1;
        while (l < r) {
            int mid = l + (r - l) / 2;
            if (nums[mid] > nums[r]) {
                l = mid + 1;  
            } else {
                r = mid;      
            }
        }
        return l; 
    }

    int bs(int l, int r, vector<int>& nums, int target) {
        while (l <= r) {
            int mid = l + (r - l) / 2;
            if (nums[mid] == target) return mid;
            else if (nums[mid] < target) l = mid + 1;
            else r = mid - 1;
        }
        return -1;
    }
    
    int search(vector<int>& nums, int target) {
        int n = nums.size();
        int pivot = findPivot(nums);

        
        if (target >= nums[pivot] && target <= nums[n - 1]) {
            return bs(pivot, n - 1, nums, target);
        } else {
            return bs(0, pivot - 1, nums, target);
        }
    }
};

File: 33-SearchinRotatedSortedArray/33-SearchinRotatedSortedArray.java
// Last updated: 8/8/2026, 10:08:35 PM
1class Solution {
2    static int bs(int[] nums, int target, int l, int r) {
3        while(l <=r) {
4            int mid = l + (r - l) / 2;
5
6            if(nums[mid] == target) return mid;
7            else if(nums[mid] < target) {
8                l = mid + 1;
9            } else {
10                r = mid - 1;
11            }
12        }
13        return -1;
14    }
15    public int search(int[] nums, int target) {
16        int l = 0;
17        int r = nums.length - 1;
18
19        while(l < r) {
20            int mid = l + (r - l) / 2;
21
22            if(nums[mid] > nums[r]) {
23                l = mid + 1;
24            } else {
25                r = mid;
26            }
27        }
28
29        int p = r;
30        l = 0;
31        r = nums.length - 1;
32        int ans = -1;
33        if(target >= nums[p] && target <= nums[r]) {
34            ans = bs(nums, target, p, r);
35        } else {
36            ans = bs(nums, target, 0, p);
37        }
38        return ans;
39    }
40}